### PR TITLE
Removed most usage of innerHTML

### LINKED ISSUE #824: Remove usage of innerHTML of our code
It is currently the case in app.js and uiUtil.js, but it should be searched everywhere (excluding the third-party libraries like fontawesome or jquery).

`.innerHTML` can be unsafe in terms of security. Its usage triggers warnings (at least when we upload the package on Mozilla store).

It should be possible to replace them by `.textContent` when it's unformatted text.
When it's not simple text, it should be possible to create the nodes programmatically.

### PR DESCRIPTION
**Pull request by aritroCoder**
Removed innerHTML from the app.js file. All features are working fine and it passed the tests.

Fixes #824 

Changes proposed: Removed usage of `innerHTML`

### CODE REVIEW

File: www/js/app.js [hunk 1]
```diff
@@ -167,7 +167,7 @@ define(['jquery', 'zimArchiveLoader', 'uiUtil', 'settingsStore','abstractFilesys
     document.getElementById('openExternalLinksInNewTabsCheck').checked = params.openExternalLinksInNewTabs;
     switchHomeKeyToFocusSearchBar();
     document.getElementById('bypassAppCacheCheck').checked = !params.appCache;
-    document.getElementById('appVersion').innerHTML = 'Kiwix ' + params.appVersion;
+    document.getElementById('appVersion').appendChild(document.createTextNode('Kiwix ' + params.appVersion));
```
Comment (mossroy) [inline www/js/app.js:170]: This works, but why did you use a different way for this one?

File: www/js/app.js [hunk 2]
```diff
@@ -622,7 +622,8 @@ define(['jquery', 'zimArchiveLoader', 'uiUtil', 'settingsStore','abstractFilesys
             apiName += ' [&nbsp;' + params.decompressorAPI.decompressorLastUsed + '&nbsp;]';
         }
         apiName = params.decompressorAPI.errorStatus || apiName || 'Not initialized';
-        decompAPIStatusDiv.innerHTML = 'Decompressor API: ' + apiName;
+        // decompAPIStatusDiv.appendChild(document.createTextNode('Decompressor API: ' + apiName));
```
Comment (mossroy) [inline www/js/app.js:625]: This commented code will need to be removed when you're done

**Review by mossroy — CHANGES_REQUESTED** [review]
This does not work in jQuery mode. How did you test?
And why did you modify package.json file?

File: www/js/app.js (hunk 1 above)
Comment (aritroCoder) [reply www/js/app.js:170]: I googled and it was showing that appenchild was one of the alternatives to using innerHTML so I did this

**Comment by aritroCoder** [comment]
> This does not work in jQuery mode. How did you test? And why did you modify package.json file?

Yes, probably it was showing correctly in my PC that time due to caching or whatever. The thing is that to render the provided HTML in the article page, we need to use innerHTML because other alternatives just displays the text. After searching some articles, I found out about this library called DOMpurify (https://github.com/cure53/DOMPurify) which filters the given HTML string for malicious codes. But I am having some issues to import the js files, I was initially trying to copy the source code  and map into the `\www\js\lib` folder, and importing it into the `index.html` using `<script type="text/javascript" src="js/lib/purify.min.js"></script>`, but when I am running `DOMPurify.sanitize(htmlArticle);` javascript is unable to recognize DOMPurify. 
Can I use npm here and import statements like server side js or is there any workarounds for this?

**Comment by mossroy** [comment]
> Yes, probably it was showing correctly in my PC that time due to caching or whatever.

OK, so please follow the CONTRIBUTING.md guidelines

> The thing is that to render the provided HTML in the article page, we need to use innerHTML because other alternatives just displays the text.

That's true. It means `.innerHTML` must be kept here.

> After searching some articles, I found out about this library called DOMpurify (https://github.com/cure53/DOMPurify) which filters the given HTML string for malicious codes. (...)

I don't see the need to use this library here.

**Comment by aritroCoder** [comment]
> > Yes, probably it was showing correctly in my PC that time due to caching or whatever.
> 
> OK, so please follow the CONTRIBUTING.md guidelines
> 
> > The thing is that to render the provided HTML in the article page, we need to use innerHTML because other alternatives just displays the text.
> 
> That's true. It means `.innerHTML` must be kept here.
> 
> > After searching some articles, I found out about this library called DOMpurify (https://github.com/cure53/DOMPurify) which filters the given HTML string for malicious codes. (...)
> 
> I don't see the need to use this library here.

Okay so I will be making a PR by keeping the innerHTML tag for that line

**Review by mossroy — CHANGES_REQUESTED** [review]
You're getting close to the goal

**Comment by Jaifroid** [comment]
@aritroCoder Are you intending to complete this PR?

**Comment by Jaifroid** [comment]
It seems this PR has been abandoned.

**Comment by aritroCoder** [comment]
Hello, I was having my exams, so I could not respond in time, I will complete this PR now please reopen it

**Comment by Jaifroid** [comment]
Oh, OK, thanks. I hope the exams went well!

File: www/js/app.js (hunk 1 above)
Comment (aritroCoder) [reply www/js/app.js:170]: > This works, but why did you use a different way for this one?

Okay, I changed it like the other areas where I used textContent and it works
I was experimenting before to find out which one of them worksso this one was different

**Comment by aritroCoder** [comment]
I have created an updated PR. Hope this goes through!

**Comment by aritroCoder** [comment]
@Jaifroid @mossroy please look through the PR, I am free now and can contribute actively if anything is required

**Comment by aritroCoder** [comment]
> @aritroCoder Thank you for this PR. The code looks good though I haven't tested yet. However, I see lots more cases of usage of `innerHTML` that could be replaced with `textContent` in both `app.js` and `uiUtil.js`. For example, take a look around line 637 and onwards in `app.js`, and around line 300 in `uiUtil.js`. Basically, you should go through every usage of `.innerHTML` and decide if it can be converted or not. If it can't be converted, because it receives actual HTML rather than text, then you should add a comment like `// Use of .innerHTML required`. There are some cases where an alternative method _might_ allow you to insert HTML, but I think for this PR it's enough that you convert all nodes that are only receiving text. I think this would not take you very long at all, the main thing is to be very careful not to convert anything that receives HTML-formatted text.

There is no removable innerHTML present in both files `app.js` and `uiUtils.js` now, I have changed them all.

**Comment by aritroCoder** [comment]
@Jaifroid Did you check it?

**Comment by Jaifroid** [comment]
I'll do so now!

**Review by Jaifroid — CHANGES_REQUESTED** [review]
@aritroCoder It looks like you need to rebase this PR on the latest version of master, as there are commits to master since you made the PR. This may be the reason for the large changed block below -- either you are inadvertently reverting the changes on master, or else you have reformatted a large block of code by mistake.

Additionally, I can still find several` .innerHTML` statements that could be changed throughout the codebase. Just as an example, look for `messageChannelStatus.innerHTML` in `app.js`. NB These are not the only ones, there are others. You need to go through them all to complete this PR!

File: www/index.html [hunk 3]
```diff
@@ -77,7 +77,7 @@ <h5 class="modal-title" id="modalLabel">Modal title goes here</h5>
         <section id="search-article" role="region">
             <header id="top">
                 <nav class="navbar navbar-expand-md bg-light" role="navigation">
-                    <a id="appVersion" class="navbar-brand">Kiwix</a>
+                    <a id="appVersion" class="navbar-brand"></a>
```
[the remark this answers is not in the document]
Comment (aritroCoder) [reply www/index.html:80]: I didnt notice that it was removed, I am adding it back

File: www/js/app.js [hunk 4]
```diff
@@ -1726,68 +1726,78 @@ define(['jquery', 'zimArchiveLoader', 'uiUtil', 'settingsStore','abstractFilesys
...
+                focusPrefixOnHomeKey
+              );
+            // when unloaded remove eventListener to avoid memory leaks
+            iframeArticleContent.contentWindow.onunload = function () {
+              iframeArticleContent.contentWindow.removeEventListener(
+                "keydown",
+                focusPrefixOnHomeKey
+              );
+            };
+          }
```
[the remark this answers is not in the document]
Comment (aritroCoder) [reply www/js/app.js:1800]: It was due to the autoformatter, I am reverting the formatting of this code

**Comment by aritroCoder** [comment]
> @aritroCoder It looks like you need to rebase this PR on the latest version of master, as there are commits to master since you made the PR. This may be the reason for the large changed block below -- either you are inadvertently reverting the changes on master, or else you have reformatted a large block of code by mistake.
> 
> Additionally, I can still find several` .innerHTML` statements that could be changed throughout the codebase. Just as an example, look for `messageChannelStatus.innerHTML` in `app.js`. NB These are not the only ones, there are others. You need to go through them all to complete this PR!

Sorry, I dont know how these got missed I was doing a CTRL+F search for `.innerHTML` . Anyways, I have removed all of them now

File: www/js/app.js [hunk 5]
```diff
@@ -169,7 +169,7 @@ define(['jquery', 'zimArchiveLoader', 'uiUtil', 'settingsStore','abstractFilesys
     document.getElementById('openExternalLinksInNewTabsCheck').checked = params.openExternalLinksInNewTabs;
     switchHomeKeyToFocusSearchBar();
     document.getElementById('bypassAppCacheCheck').checked = !params.appCache;
-    document.getElementById('appVersion').innerHTML = 'Kiwix ' + params.appVersion;
+    document.getElementById('appVersion').textContent = 'Kiwix ' + params.appVersion;
```
Comment (Jaifroid) [inline www/js/app.js:172]: Please remove extraneous space at the end of this line.

File: www/js/app.js [hunk 6]
```diff
@@ -634,36 +634,36 @@ define(['jquery', 'zimArchiveLoader', 'uiUtil', 'settingsStore','abstractFilesys
...
         var serviceWorkerStatus = document.getElementById('serviceWorkerStatus');
         if (isMessageChannelAvailable()) {
-            messageChannelStatus.innerHTML = 'MessageChannel API available';
+            messageChannelStatus.textContent = 'MessageChannel API available';
             messageChannelStatus.classList.remove('apiAvailable', 'apiUnavailable');
             messageChannelStatus.classList.add('apiAvailable');
         } else {
             apiPanelClass = 'card-warning';
-            messageChannelStatus.innerHTML = 'MessageChannel API unavailable';
+            messageChannelStatus.textContent = "MessageChannel API unavailable";
```
Comment (Jaifroid) [inline www/js/app.js:642]: Why did you change the quotation style? This is explicitly mentioned in CONTRIBUTING.md... You should keep the quotation style of the area you are editing, for consistency.

File: www/js/app.js [hunk 7]
```diff
@@ -634,36 +634,36 @@ define(['jquery', 'zimArchiveLoader', 'uiUtil', 'settingsStore','abstractFilesys
...
                 apiPanelClass = 'card-warning';
-                serviceWorkerStatus.innerHTML = 'ServiceWorker API available, but not registered';
+                serviceWorkerStatus.textContent = 'ServiceWorker API available, but not registered';
                 serviceWorkerStatus.classList.remove('apiAvailable', 'apiUnavailable');
                 serviceWorkerStatus.classList.add('apiUnavailable');
             }
         } else {
             apiPanelClass = 'card-warning';
-            serviceWorkerStatus.innerHTML = 'ServiceWorker API unavailable';
+            serviceWorkerStatus.textContent = "ServiceWorker API unavailable";
```
Comment (Jaifroid) [inline www/js/app.js:659]: Ditto.

File: www/js/lib/uiUtil.js [hunk 8]
```diff
@@ -225,7 +225,7 @@ define(rqDef, function(settingsStore) {
             if (typeof URL === 'function') return new URL(url, base);
             // IE11 lacks URL API: workaround adapted from https://stackoverflow.com/a/28183162/9727685
             var d = document.implementation.createHTMLDocument('t');
-            d.head.innerHTML = '<base href="' + base + '">';
+            d.head.textContent = '<base href="' + base + '">';
```
Comment (Jaifroid) [inline www/js/lib/uiUtil.js:228]: This is an HTML document. You can't use textContent here.

File: www/js/lib/uiUtil.js [hunk 9]
```diff
@@ -302,9 +302,9 @@ define(rqDef, function(settingsStore) {
         a.type = contentType;
         a.download = filename;
         a.classList.add('alert-link');
-        a.innerHTML = filename;
+        a.textContent = filename;
         var alertMessage = document.getElementById('alertMessage');
-        alertMessage.innerHTML = '<strong>Download</strong> If the download does not start, please tap the following link: ';
+        alertMessage.textContent = '<strong>Download</strong> If the download does not start, please tap the following link: ';
```
Comment (Jaifroid) [inline www/js/lib/uiUtil.js:307]: There are HTML tags here. You can't use textContent.

**Review by Jaifroid — CHANGES_REQUESTED** [review]
Thanks for this PR, @aritroCoder. I think you got all the innerHTML that can be changed now. However, there are some that you¡'ve changed which I don't think can be changed. See individual comments below. There are also some minor formatting changes: although it may appear unimportant, it's a good idea to ensure your changes are minimal or are an improvement when you make PRs, so please change these too as indicated in the comments. Once you've done this, I'll do some testing and then if all is well, I'll be happy to approve!

**Comment by Jaifroid** [comment]
PS You could change this PR title to "Removed most usage of innerHTML". If it doesn't let you change it, no worries, we can simply set the name of the commit at squash/merge time.

**Comment by aritroCoder** [comment]
> Thanks for this PR, @aritroCoder. I think you got all the innerHTML that can be changed now. However, there are some that you¡'ve changed which I don't think can be changed. See individual comments below. There are also some minor formatting changes: although it may appear unimportant, it's a good idea to ensure your changes are minimal or are an improvement when you make PRs, so please change these too as indicated in the comments. Once you've done this, I'll do some testing and then if all is well, I'll be happy to approve!

Done, now you can test the PR

**Comment by aritroCoder** [comment]
@Jaifroid @mossroy did you test the PR?

**Review by Jaifroid — APPROVED** [review]
@aritroCoder Apologies for the delay getting back to you. Everything looks good now, and my own testing has passed. Many thanks for this PR. I'll squash/merge later today unless you tell me otherwise. It will appear as your contribution/commit.

**Comment by aritroCoder** [comment]
> @aritroCoder Apologies for the delay getting back to you. Everything looks good now, and my own testing has passed. Many thanks for this PR. I'll squash/merge later today unless you tell me otherwise. It will appear as your contribution/commit.

Welcome! I am really happy to contribute
One final question, I wish to contribute here to participate in GSoC 23, so is there any channel/forum for gsoc preperation and for participants here?

**Comment by Jaifroid** [comment]
@aritroCoder It's good to know you're interested in GSoC 23. I'm pinging @kelson42 because participation is at Kiwix Org level rather than per-Repo. You could start by joining the [Kiwix Slack channel](https://kiwixoffline.slack.com/) if you aren't already a member. @kelson42, anything else? Will Kiwix be applying to be an approved Org this year (if it works like that)?
